src/mdns/mdns_mdnssd.cpp

```diff
@@ -313,67 +349,117 @@ void PublisherMDnsSd::HandleServiceRegisterResult(DNSServiceRef         aService
                                                   const char *          aType,
                                                   const char *          aDomain)
 {
-    otbrLog(OTBR_LOG_INFO, "Got a reply for service %s.%s%s", aName, aType, aDomain);
+    OTBR_UNUSED_VARIABLE(aDomain);
+
+    otbrError       error = DNSErrorToOtbrError(aError);
+    std::string     originalInstanceName;
+    ServiceIterator service = FindPublishedService(aServiceRef);
+
+    otbrLog(OTBR_LOG_INFO, "[mdns] got a reply for service %s.%s", aName, aType);
+
+    if (service != mServices.end())
+    {
+        originalInstanceName = service->mName;
+    }
 
     if (aError == kDNSServiceErr_NoError)
     {
         if (aFlags & kDNSServiceFlagsAdd)
         {
-            otbrLog(OTBR_LOG_INFO, "MDNS added service %s", aName);
             RecordService(aName, aType, aServiceRef);
         }
         else
         {
-            otbrLog(OTBR_LOG_INFO, "MDNS remove service %s", aName);
             DiscardService(aName, aType, aServiceRef);
         }
     }
     else
     {
-        otbrLog(OTBR_LOG_ERR, "Failed to register service %s: %s", aName, DNSErrorToString(aError));
+        otbrLog(OTBR_LOG_ERR, "[mdns] failed to register service %s: %s", aName, DNSErrorToString(aError));
         DiscardService(aName, aType, aServiceRef);
     }
+
+    if (mServiceHandler != nullptr)
+    {
+        // mDNSResponder could auto-rename the service instance name when name conflict
+        // is detected. In this case, `aName` may don't match `service->mName` and we
+        // should use the original `service->mName` to find associated SRP service.
+        // TODO: pass the re-newed service instance name back to SRP server handler.
+        const char *name = !originalInstanceName.empty() ? originalInstanceName.c_str() : aName;
+
+        mServiceHandler(name, service->mType, error, mServiceHandlerContext);
```

Comment:
In case service is discarded, better to use `aType` instead.
```suggestion
        mServiceHandler(name, aType, error, mServiceHandlerContext);
```